File: frontend/src/app/core/services/billing.service.ts
import { Injectable } from "@angular/core";
import { HttpClient, HttpParams } from "@angular/common/http";
import { Observable } from "rxjs";
import { environment } from "../../../environments/environment";
import { Invoice, ReportSummary } from "../../shared/models/invoice.model";

@Injectable({ providedIn: "root" })
export class BillingService {
  private base = environment.apiUrl;
  constructor(private http: HttpClient) {}

  // ── Invoices ─────────────────────────────────────────────────
  getInvoices(params?: any): Observable<Invoice[]> {
    return this.http.get<Invoice[]>(`${this.base}/invoices`, { params });
  }
  getInvoice(id: number): Observable<Invoice> {
    return this.http.get<Invoice>(`${this.base}/invoices/${id}`);
  }
  createInvoice(req: any): Observable<Invoice> {
    return this.http.post<Invoice>(`${this.base}/invoices`, req);
  }
  updateInvoice(id: number, req: any): Observable<Invoice> {
    return this.http.put<Invoice>(`${this.base}/invoices/${id}`, req);
  }
  voidInvoice(id: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/invoices/${id}`);
  }
  generateFromAppointment(apptId: number): Observable<Invoice> {
    return this.http.post<Invoice>(`${this.base}/invoices/from-appointment/${apptId}`, {});
  }

  // ── Payments ─────────────────────────────────────────────────
  getPayments(params?: any): Observable<any[]> {
    return this.http.get<any[]>(`${this.base}/payments`, { params });
  }
  postPayment(req: any): Observable<any> {
    return this.http.post<any>(`${this.base}/payments`, req);
  }

  // ── Reports ──────────────────────────────────────────────────
  getDailyReport(date: string): Observable<ReportSummary> {
    return this.http.get<ReportSummary>(`${this.base}/reports/daily`,
      { params: new HttpParams().set("date", date) });
  }
  getMonthlyReport(year: number, month: number): Observable<ReportSummary> {
    return this.http.get<ReportSummary>(`${this.base}/reports/monthly`,
      { params: new HttpParams().set("year", String(year)).set("month", String(month)) });
  }
  getYtdReport(year: number): Observable<ReportSummary> {
    return this.http.get<ReportSummary>(`${this.base}/reports/ytd`,
      { params: new HttpParams().set("year", String(year)) });
  }
}

File: backend/src/main/java/com/yourowncrm/service/BillingService.java
package com.yourowncrm.service;

import com.yourowncrm.dto.request.InvoiceRequest;
import com.yourowncrm.dto.request.PaymentRequest;
import com.yourowncrm.dto.response.InvoiceResponse;
import com.yourowncrm.dto.response.ReportSummary;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

public interface BillingService {
    InvoiceResponse  generateInvoiceFromAppointment(UUID tenantId, Long appointmentId, Long userId);
    InvoiceResponse  createInvoice(UUID tenantId, InvoiceRequest req, Long userId);
    InvoiceResponse  updateInvoice(UUID tenantId, Long invoiceId, InvoiceRequest req, Long userId);
    InvoiceResponse  voidInvoice(UUID tenantId, Long invoiceId, Long userId);
    InvoiceResponse  getInvoice(UUID tenantId, Long invoiceId);
    List<InvoiceResponse> getInvoices(UUID tenantId, LocalDate from, LocalDate to);

    void             postPayment(UUID tenantId, PaymentRequest req, Long userId);

    ReportSummary    getDailyReport(UUID tenantId, LocalDate date);
    ReportSummary    getMonthlyReport(UUID tenantId, int year, int month);
    ReportSummary    getYtdReport(UUID tenantId, int year);
}
